lib/puppet/provider/rhsm_reconfigure_script/rhsm_reconfigure_script.rb

```diff
@@ -0,0 +1,151 @@
+Puppet::Type.type(:rhsm_reconfigure_script).provide(:rhsm_reconfigure_script) do
+
+  def create
+    template = render_template
+    File.open(resource[:name], 'w', 0o755) do |file|
+      file.write(template)
+    end
+  end
+
+  def exists?
+    File.exist?(resource[:name]) && (File.read(resource[:name]) == render_template)
+  end
+
+  private
+
+  def render_template
+    ::ERB.new(script_template).result(binding)
+  end
+
+  def script_template
+    <<~'HEREDOC'
+      #!/bin/bash
+
+      set -e
+
+      KATELLO_SERVER=<%= resource[:rhsm_hostname] %>
+      PORT=<%= resource[:rhsm_port] %>
+
+      KATELLO_SERVER_CA_CERT=<%= resource[:server_ca_name] %>.pem
+      KATELLO_DEFAULT_CA_CERT=<%= resource[:default_ca_name] %>.pem
+
+      CERT_DIR=/etc/rhsm/ca
+      PREFIX=/rhsm
+      CFG=/etc/rhsm/rhsm.conf
+      CFG_BACKUP=$CFG.kat-backup
+      CA_TRUST_ANCHORS=/etc/pki/ca-trust/source/anchors
+
+      read -r -d '' KATELLO_DEFAULT_CA_DATA << EOM || true
+      <%= File.read(resource[:default_ca_cert]) %>
```

Comment:
I thought I always understood it that providers were executed on clients as they acted like any other resource (e.g. file, service, exec)

Reply:
I think you may be right indeed. I still feel that a helper method with a comment can explain why this is a type + provider rather than a file + template. Normally I'm minimalistic in my comments but when something is unusual it helps to highlight what part is unusual.